File: pkm/widgets/nowplaying.py
import dbus, os, random, requests
from shutil import copyfile
from pkm.widgets.base import BaseWidget
from pkm import CACHE, PKMETER, ROOT
from pkm import log, utils

DEFAULT_ART = f"{ROOT}/pkm/img/nowplaying.jpg"
NOWPLAYING_ART = f"{CACHE}/nowplaying/nowplaying.jpg"
EMPTY_DATA = {'application':'', 'status':'', 'title':'Current Track', 'artist':'Artist', 'arturl':'', 'length':'0:00', 'position':'0:00', 'percent':0}
QUIPS = ["All is calm", "Enjoying the peace and quiet", "I can hear my thoughts", "It's quiet in here",
    "No beats detected", "No jams right now", "No melodies at the moment", "No music detected",
    "No music in the queue", "No music playing", "No music, no problem", "No rhythm in the air",
    "No songs in the queue", "No sound waves detected", "No tracks available", "No tracks spinning",
    "No tunes at the moment", "No tunes available", "No tunes playing", "No tunes to groove to",
    "Nothing playing right now", "Quiet as a mouse", "Shhh...it's quiet time", "Silence is golden",
    "The audio is on hold", "The band is on a break", "The beats are on break", "The concert is paused",
    "The dance floor is empty", "The DJ is taking a nap", "The jukebox is silent",
    "The music has gone fishing", "The music has left the building", "The music has taken a break",
    "The music is off-duty", "The music is on a break", "The music is on break", "The music is on hiatus",
    "The music is on hold", "The music is on pause", "The music is on standby", "The music is on standby",
    "The music is on vacation", "The music is taking a nap", "The orchestra is tuning up",
    "The playlist is empty", "The playlist is on break", "The playlist is on hold",
    "The playlist is on vacation", "The playlist is paused", "The playlist is silent",
    "The playlist is taking five", "The sound is on sabbatical", "The sound of silence",
    "The sound system is idle", "The sound system is idle", "The sound system is resting",
    "The sound system is silent", "The soundscape is barren", "The speakers are off",
    "The speakers are on mute", "The speakers are resting", "The stage is empty"]


class NowPlayingWidget(BaseWidget):
    """ Displays the NVIDIA gpu metrics. """

    def __init__(self, wsettings, origin=0):
        super().__init__(wsettings, origin)
        # Header and footer are 67px
        # Each row of Ubuntu:size=8 adds 13px height
        self.height = 111

    def get_conkyrc(self, theme):
        """ Create the conkyrc template for the this widget. """
        return utils.clean_spaces(f"""
            ${{texeci {self.update_interval} {PKMETER} update {self.name}}}\\
            ${{voffset 20}}${{goto 10}}{theme.header}Now Playing
            ${{goto 10}}{theme.subheader}${{execi 2 {PKMETER} get {self.name}.application}}
            ${{voffset 22}}${{goto 10}}{theme.value}${{execi 2 {PKMETER} get {self.name}.title | cut -c 1-23}}
            ${{goto 10}}{theme.label}${{execi 2 {PKMETER} get {self.name}.artist | cut -c 1-23}}
            ${{goto 10}}{theme.label}${{execi 2 {PKMETER} get {self.name}.position}} of ${{execi 5 {PKMETER} get {self.name}.length}}
        """)  # noqa

    def get_lua_entries(self, theme):
        """ Create the draw.lua entries for this widget. """
        origin, width, height = self.origin, self.width, self.height
        pct = f'execi 2 {PKMETER} get {self.name}.percent'
        return [
            self.draw('line', frm=(100,origin), to=(100,origin+40), thickness=width, color=theme.header_bg),  # header bg
            self.draw('line', frm=(100,origin+40), to=(100,origin+height), thickness=width, color=theme.bg),  # main bg
            self.draw('bar_graph', conky_value=pct, frm=(10,origin+50), to=(130,origin+50), bar_color=theme.accent1,
                bar_thickness=2, background_color=theme.graph_bg, critical_threshold=105),  # percent playing
            self.draw('image', filepath=NOWPLAYING_ART, frm=(140,origin+48), width=50),  # album art
        ]

    def update_cache(self):
        """ Fetch current track information from dbus and update cache. """
        newdata = {}
        olddata = utils.load_cached_data(self.cachepath) or EMPTY_DATA
        # Fetch the track information from dbus
        session = dbus.SessionBus()
        players = (session.get_object('org.freedesktop.DBus', '/org/freedesktop/DBus')
            .ListNames(dbus_interface='org.freedesktop.DBus'))
        players = [name for name in players if name.startswith('org.mpris.MediaPlayer2.')]
        for player in players:
            try:
                playerobj = session.get_object(player, '/org/mpris/MediaPlayer2')
                properties = dbus.Interface(playerobj, 'org.freedesktop.DBus.Properties')
                metadata = properties.Get('org.mpris.MediaPlayer2.Player', 'Metadata')
                newdata['status'] = properties.Get('org.mpris.MediaPlayer2.Player', 'PlaybackStatus')
                newdata['title'] = str(metadata.get('xesam:title', 'Unknown Title'))
                if newdata['status'] == 'Playing' and newdata['title'] != 'Unknown':
                    newdata['application'] = str(properties.Get('org.mpris.MediaPlayer2', 'Identity'))
                    newdata['artist'] = str(metadata.get('xesam:artist', ['Unknown Artist'])[0])
                    newdata['length'] = metadata.get('mpris:length', 0) // 1000000
                    newdata['position'] = properties.Get('org.mpris.MediaPlayer2.Player', 'Position') // 1000000
                    newdata['percent'] = utils.percent(newdata['position'], newdata['length'])
                    newdata['position'] = self._format_duration(newdata['position'])
                    newdata['length'] = self._format_duration(newdata['length'])
                    newdata['arturl'] = str(metadata.get('mpris:artUrl', ''))
                    self._download_art(newdata['arturl'])
                    return newdata
            except Exception as err:
                log.error(f"Error fetching player info: {err}")
        # No track information was found
        copyfile(DEFAULT_ART, NOWPLAYING_ART)
        newdata = dict(**EMPTY_DATA)
        quip = olddata['application']
        newdata['application'] = quip if quip in QUIPS else random.choice(QUIPS)
        return newdata
        
    def _download_art(self, url):
        """ Download album art to cache and return path. """
        if not url: return None
        os.makedirs(f'{CACHE}/{self.name}', exist_ok=True)
        log.info(f"Downloading album art from {url}")
        response = requests.get(url, stream=True)
        if response.status_code == 200:
            with open(NOWPLAYING_ART, 'wb') as handle:
                handle.write(response.content)

    def _format_duration(self, seconds):
        """Convert seconds to a string in the format n:nn."""
        minutes, seconds = divmod(seconds, 60)
        return f'{minutes}:{seconds:02d}'

File: pkm/widgets/nowplaying.lua
local config = require 'config'
local draw = require 'pkm/draw'
local utils = require 'pkm/utils'

local nowplaying = {}
nowplaying.PLAYERCTL_FORMAT = {'playername={{playerName}}', 'status={{status}}',
  'title={{title}}', 'artist={{artist}}', 'album={{album}}', 'length={{mpris:length}}',
  'position={{position}}', 'arturl={{mpris:artUrl}}'}
nowplaying.origin = 0
nowplaying.height = 0
nowplaying.players = {}
nowplaying.last_update = nil

-- Draw
-- Draw this widget
function nowplaying:draw()
  if #self.players == 0 then
    self.height = 0
    return
  end
  self.height = 51

  -- Header
  local playernames = self:list_playernames(self.players, self.max_players)
  draw.rectangle{x=0, y=self.origin+0, width=conky_window.width, height=40, color=config.header_bg} -- header background
  draw.rectangle{x=0, y=self.origin+40, width=conky_window.width, height=self.height-40, color=config.background} -- background
  draw.text{x=10, y=self.origin+17, text='Now Playing', size=12, color=config.header} -- now playing
  draw.text{x=10, y=self.origin+32, text=playernames, color=config.subheader, maxwidth=180} -- player names
  
  -- Play Buttons
  local imgnext = pkmeter.ROOT..'/pkm/img/nowplaying/next.png'
  local imgplay = pkmeter.ROOT..'/pkm/img/nowplaying/play.png'
  local imgpause = pkmeter.ROOT..'/pkm/img/nowplaying/pause.png'
  local imgprev = pkmeter.ROOT..'/pkm/img/nowplaying/prev.png'
  draw.image{x=122, y=y+6, path=imgprev, width=21}
  if self.players[1].status == 'Playing' then
    draw.image{x=147, y=y+6, path=imgpause, width=21}
  else
    draw.image{x=147, y=y+6, path=imgplay, width=21}
  end
  draw.image{x=172, y=y+6, path=imgnext, width=21}
  

  -- Player Info
  local y = self.origin + 50
  for _, player in ipairs(self.players) do
    local duration = utils.duration(player.position)..' of '..utils.duration(player.length)
    draw.rectangle{x=0, y=y-1, width=conky_window.width, height=67, color=config.background} -- background
    draw.bargraph{x=10, y=y, width=120, height=2, value=player.position, maxvalue=player.length,
      color=config.accent, bgcolor=config.graph_bg} -- progress
    draw.text{x=10, y=y+17, text=player.title, maxwidth=120, color=config.value} -- title
    draw.text{x=10, y=y+32, text=player.artist, maxwidth=120, color=config.label} -- artist
    draw.text{x=10, y=y+47, text=duration, color=config.label} -- duration
    draw.image{x=140, y=y+0, path=player.artpath, width=50}
    self.height = self.height + 67
    y = y + 67
  end
end

-- Update
-- Update Playerctl info
function nowplaying:update()
  if utils.check_update(self.last_update, config.update_interval) then
    players = {}
    cmd = self.playerctl..' metadata -af "'..table.concat(self.PLAYERCTL_FORMAT, ';;')..'"'
    if self.ignore_players and #self.ignore_players > 0 then
      cmd = cmd..' -i '..self.ignore_players
    end
    local result = utils.run_command(cmd)
    i = 1
    for line in string.gmatch(result, "[^\r\n]+") do
      local player = {}
      for keyval in string.gmatch(line, "([^;;]+)") do
        local key, val = string.match(keyval, "^(%S+)=(.+)$")
        if key and val then player[key] = val end
      end
      -- Cleanup length & position, download art
      if player.length then player.length = math.floor(tonumber(player.length) / 1000000) end
      if player.position then player.position = math.floor(tonumber(player.position) / 1000000) end
      if player.arturl then player.artpath = self:get_artpath(player.playername, player.arturl) end
      table.insert(players, player)
      if #players >= self.max_players then break end
      i = i + 1
    end
    -- Clenaup old _art objects (in case it leaks memory)
    for _, player in ipairs(self.players) do player._art = nil end
    self.players = players
    self.last_update = os.time()
  end
end

-- Click
-- Perform click action
function nowplaying:click(event, x, y)
  if 5 <= y and y <= 5+20 then
    if 122 <= x and x <= 122+20 then
      os.execute(self.playerctl..' previous')
    elseif 147 <= x and x <= 147+20 then
      os.execute(self.playerctl..' play-pause')
    elseif 172 <= x and x <= 172+20 then
      os.execute(self.playerctl..' next')
    end
  end
end

-- Get Image Data
-- Check we need to download the image for this player
function nowplaying:get_artpath(playername, arturl)
  local player = self:get_player_by_name(playername)
  if player and player.arturl == arturl then
    return player.artpath
  elseif string.sub(arturl, 1, 7) == 'file://' then
    return string.sub(arturl, 7)
  else
    return utils.download_image(arturl)
  end
end

-- Get Player by Name
-- Returns the player object by playername
function nowplaying:get_player_by_name(playername)
  for _, player in ipairs(self.players) do
    if player.playername == playername then return player end
  end
  return nil
end

-- List Playernames
-- Returns a comma seperated string of player names
function nowplaying:list_playernames(players, max_players)
  local names = {}
  for _, player in ipairs(players) do
    local name = utils.titleize(player.playername)
    table.insert(names, name)
  end
  return table.concat(names, ', ')
end

return nowplaying